ui/src/middlewares/searchDispatcher.js

```diff
@@ -1,68 +1,109 @@
-import { LOCATION_CHANGE } from 'connected-react-router';
 import omit from 'lodash.omit';
 import {
-  searchForCurrentLocation,
-  fetchSearchAggregationsForCurrentLocation,
+  searchForCurrentQuery,
+  fetchSearchAggregationsForCurrentQuery,
+  newSearch,
 } from '../actions/search';
-import { LITERATURE, JOBS, AUTHORS } from '../common/routes';
 import { shallowEqual } from '../common/utils';
+import {
+  SEARCH_BASE_QUERIES_UPDATE,
+  SEARCH_QUERY_UPDATE,
+} from '../actions/actionTypes';
 
-function getLocationFromRootState(state) {
-  const {
-    router: { location },
-  } = state;
-  return location || {};
+function getQueryFromState(namespace, state) {
+  return state.search.getIn(['namespaces', namespace, 'query']);
 }
 
-const COLLECTION_PATHNAMES = [LITERATURE, JOBS, AUTHORS];
-const idInUrlRegExp = new RegExp('/\\d+');
-function isSearchPage(location) {
-  const isCollectionPage = COLLECTION_PATHNAMES.some(pathname =>
-    location.pathname.startsWith(pathname)
-  );
+function getBaseQueryFromState(namespace, state) {
+  return state.search.getIn(['namespaces', namespace, 'baseQuery']);
+}
+
+function getBaseAggregationsQueryFromState(namespace, state) {
+  return state.search.getIn(['namespaces', namespace, 'baseAggregationsQuery']);
+}
 
-  return isCollectionPage && !idInUrlRegExp.test(location.pathname);
+function hasQueryChangedExceptSortOrPage(prevQuery, nextQuery) {
+  return !shallowEqual(
+    omit(prevQuery.toJS(), ['sort', 'page']),
+    omit(nextQuery.toJS(), ['sort', 'page'])
+  );
 }
 
+// TODO: this could be moved to the action creator
+// only advantage is that it's easier to test SEARCH_QUERY_UPDATE in search container components
+// rather than an async action
 export default function({ getState, dispatch }) {
   return next => action => {
-    if (
-      action.type === LOCATION_CHANGE &&
-      action.payload &&
-      isSearchPage(action.payload.location)
-    ) {
-      const currentLocation = getLocationFromRootState(getState());
-      const result = next(action);
-      const nextLocation = action.payload.location; // use this instead of `getLocationFromRootState(getState())` for testability.
-      const { isFirstRendering } = action.payload;
-      const hasPathnameChanged =
-        nextLocation.pathname !== currentLocation.pathname;
-      const hasSearchChanged = nextLocation.search !== currentLocation.search;
-      if (isFirstRendering || hasPathnameChanged || hasSearchChanged) {
-        dispatch(searchForCurrentLocation());
+    if (action.type === SEARCH_QUERY_UPDATE) {
+      const { namespace, query } = action.payload;
+
+      const stateQuery = getQueryFromState(namespace, getState());
+      const hasQueryQParamChanged = query.q && query.q !== stateQuery.get('q');
+      if (hasQueryQParamChanged) {
+        dispatch(newSearch(namespace));
```

Comment:
newSearch naming was a bit confusing for me, maybe refreshSearch is better? or resetSearch?

Reply:
I agree it was like that before, if there is time I'll rename it.